poly-commit/src/hyrax/mod.rs

```diff
@@ -0,0 +1,533 @@
+///   default implementations: `batch_open`, `batch_check`,
+///   `open_combinations`, `check_combinations`. This is not discussed in the
+///   reference article, but the IPA and KZG modules might be a good starting
+///   point.
+/// - On a related note to the previous point, there might be a more
+///   efficient way to open several polynomials at a single point (this is the
+///   functionality of the `open` method) than the currently implemented
+///   technique, where only the computation of the vectors `L` and `R` is
+///   shared across polynomials.
+/// - The cited article proposes an optimisation in the section _Reducing the
+///   cost of proof-of-dot-prod_. It allows for non-square matrices (and hence
+///   removes the requirement for the number of variables to be even) and
+///   introduces a tradeoff between proof size and verifier time. It is
+///   probably worth pursuing.
+
+pub struct HyraxPC<
+    // The elliptic curve used for Pedersen commitments (only EC groups are
+    // supported as of now).
+    G: AffineRepr,
+    // A polynomial type representing multilinear polynomials
+    P: MultilinearExtension<G::ScalarField>,
+    // The sponge used in the protocol as random oracle
+    S: CryptographicSponge,
+> {
+    _phantom: PhantomData<(G, P, S)>,
+}
+
+impl<G, P, S> HyraxPC<G, P, S>
+where
+    G: AffineRepr,
+    P: MultilinearExtension<G::ScalarField>,
+    S: CryptographicSponge,
+{
+    /// Pedersen commitment to a vector of scalars as described in appendix A.1
+    /// of the reference article.
+    /// The caller must either directly pass hiding exponent `r` inside Some,
+    /// or provide an rng so that `r` can be sampled.
+    /// If there are `n` scalars, the first `n` elements of the key will be
+    /// multiplied by them in the same order, and its `n + 1`th element will be
+    /// multiplied by `r`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if both `r` and `rng` are None.
+    fn pedersen_commit(
+        key: &HyraxCommitterKey<G>,
+        scalars: &[G::ScalarField],
+        r: Option<G::ScalarField>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> (G, G::ScalarField) {
+        // Cannot use unwrap_or, since its argument is always evaluated
+        let r = match r {
+            Some(v) => v,
+            None => G::ScalarField::rand(rng.expect("Either r or rng must be provided")),
+        };
+
+        let mut scalars_ext = Vec::from(scalars);
+        scalars_ext.push(r);
+
+        // Trimming the key to the length of the coefficient vector
+        let mut points_ext = key.com_key[0..scalars.len()].to_vec();
+        points_ext.push(key.h);
+
+        let scalars_bigint = ark_std::cfg_iter!(scalars)
+            .map(|s| s.into_bigint())
+            .collect::<Vec<_>>();
+
+        // Multi-exponentiation in the group of points of the EC
+        let com = <G::Group as VariableBaseMSM>::msm_bigint(&points_ext, &scalars_bigint);
```

Comment:
This requires an extra allocation. A better way to compute this would be just do an MSM with `com_key`, and then add in the product `h * r` separately.

This latter addition can also be pulled out of the `pedersen_commit` (which means that you don't need to pass in `rng` or `r`)

Reply:
Done!

Reply:
I think there was a bug too. `r` was pushed into `scalars_ext` but `scalars_ext` was never used.